### PR TITLE
New Feature: jplsbdb 

### PR DESCRIPTION
**Pull request by mommermi**
jplsbdb is a query service for JPL's Small-Body Database Browser (https://ssd.jpl.nasa.gov/sbdb.cgi). It provides information on Solar System small bodies and re-arranges them into a dictionary.

@migueldvb, @mkelley: could you please review this?

### CODE REVIEW

**Comment by mommermi** [comment]
@bsipocz @keflavich 

I think the search engine (https://ssd.jpl.nasa.gov/sbdb_query.cgi) could be included here, as it taps the same data. They do not yet provide an API for that service, so we will have to wait a bit for this to actually happen.

I think all the jpl services are different enough that they merit individual modules. There are a few more that we could implement in the future (here is a full list: https://ssd-api.jpl.nasa.gov/), but we don't have to. You would have to tell me what the critical mass for unification is. If you prefer putting all jpl services into one big module, we should think about this early on...

**Comment by keflavich** [comment]
I would prefer hosting them in the same module, but we should have different classes.  This hasn't really come up for anything else, though in principle we could have had `cds/simbad`, `cds/vizier`, `cds/cds_otherthing` at this point.  None of the other providers provide multiple services, which is why this hasn't come up before.

We might also need to consider some global reorganization as astroquery continues to grow.  I'll raise a different issue for that.

**Comment by mommermi** [comment]
Ok, I can take care of this - shouldn't be too hard. Let me re-submit this pr in an umbrella JPL module. Would you like to have a separate `jplhorizons` with deprecation warning for a while, just to have people know what's going on? We can then remove that submodule, and only keep `jpl`.

@gianninapr @migueldvb: would you like `jplspec` to be implemented in an umbrella `jpl` module, too?

**Comment by migueldvb** [comment]
I can review this pull request in the end of the week if it is not too late!

Regarding the integration of JPL modules in an umbrella namespace for JPL-provided services, I think that `jplspec`, although we use it in `sbpy`,  is not specific to solar system science and it has been used in other fields as well. Perhaps it would be more useful for astroquery users if `jplspec` is organized together with other molecular databases like `splatalogue`, `lamda`, `cdms`... and not based on the organization that provides the service. I'll let @gianninapr chime in with her opinion about where `jplspec` should go.

**Comment by gianninapr** [comment]
I think the biggest factor to consider with ``jplspec`` is that it is not solar system science specific, as @migueldvb mentioned. For this sole reason, keeping ``jplspec`` grouped with the other molecular databases makes more sense to me. I fear the potential usefulness and goal of ``jplspec`` going unnoticed if it's under a different grouping. Although I might be overlooking some of the more structural factors, so in the end I leave it in all of your hands to decide!

**Comment by mommermi** [comment]
@gianninapr @migueldvb: I fully agree, let's keep `jplspec` out of `jpl`! 

@migueldvb: there is no rush reviewing this. I'll be working on putting everything into an umbrella module in the meantime.

File: astroquery/jplsbdb/core.py [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+import json
+
+# 2. third party imports
+from numpy import genfromtxt, isnan, array
+import astropy.units as u
+
+# 3. local imports - use relative imports
+from ..query import BaseQuery
+from ..utils import commons
+from ..utils import prepend_docstr_nosections
```
Comment (migueldvb) [inline astroquery/jplsbdb/core.py]: `..utils.prepend_docstr_nosections` and `..utils.commons` imported but unused

File: astroquery/jplsbdb/core.py [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+
+                    # leave scalar elements as they are
+                    else:
+                        res[key] = val
+
+            # re-run this function on leaf dicts
+            elif isinstance(val, dict):
+                res[key] = self._process_data(val)
+
+             # use leaf scalars (and apply units, where applicable)
```
Comment (migueldvb) [inline astroquery/jplsbdb/core.py]: over indented by one space

File: astroquery/jplsbdb/tests/test_jplsbdb.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+import pytest
+import os
+from collections import OrderedDict
```
Comment (migueldvb) [inline astroquery/jplsbdb/tests/test_jplsbdb.py]: `OrderedDict` unused

File: docs/jplsbdb/jplsbdb.rst [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+to the `Small-Body Database Browser
+<https://ssd.jpl.nasa.gov/sbdb.cgi>`_ (SBDB) maintained by the `JPL Solar
+System Dynamics group <http://ssd.jpl.nasa.gov/>`_.
+
+The SBDB provides detailed information on a specific known small body,
+including it's orbit, close approaches with major planets, available
+radar observations, detailed information on virtual impactors,
+discovery circumstances, and a few select physical properties.
+
+This module enables the query of these information for an indiviual
```
Comment (migueldvb) [inline docs/jplsbdb/jplsbdb.rst]: individual

File: astroquery/jplsbdb/core.py [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+                if 'notes' in q:
+                    eldict[q['name']+'_note'] = q['notes']
+
+        return eldict
+
+    def schematic(self, d, _prepend='+--'):
+        """
+        Formats the provided dictionary ``d`` into a human-readable tree
+        structure schematic. In order to display the structure
+        properly, the resulting ``oustring`` should be passed to
```
Comment (migueldvb) [inline astroquery/jplsbdb/core.py]: oustring -> outstring

File: astroquery/jplsbdb/core.py [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+            Return raw data as obtained by JPL SBDB without parsing the
+            data. Default: ``False``
+        get_uri : boolean, optional
+            Add the query URI to the output as ``query_uri`` field.
+            Default: ``False``
+
+        Returns
+        -------
+        res : dict
+            A dicitionary holding all the parsed data.
```
Comment (migueldvb) [inline astroquery/jplsbdb/core.py]: dictionary

File: astroquery/jplsbdb/core.py [hunk 7]
```diff
@@ -0,0 +1,416 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+# 1. standard library imports
+import json
+
+# 2. third party imports
+from numpy import genfromtxt, isnan, array
+import astropy.units as u
+
+# 3. local imports - use relative imports
```
Comment (migueldvb) [inline astroquery/jplsbdb/core.py]: I think the comments before the import statements are probably not needed

File: astroquery/jplsbdb/core.py [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+
+    def schematic(self, d, _prepend='+--'):
+        """
+        Formats the provided dictionary ``d`` into a human-readable tree
+        structure schematic. In order to display the structure
+        properly, the resulting ``oustring`` should be passed to
+        the ``print`` function.
+
+                Parameters
+        ----------
```
Comment (migueldvb) [inline astroquery/jplsbdb/core.py]: Parameters heading over indented

**Comment by mommermi** [comment]
@migueldvb: thanks for your revision! I adopted all your suggestions and replaced dictionaries with OrderedDicts wherever possible.

**Comment by mommermi** [comment]
@keflavich @bsipocz @migueldvb 
Ok, I spent some time re-organizing `jplsbdb` into `solarsystem/jpl/sbdb`. This is not yet final, just an idea for now since some issues popped up:
* in #1215, we talked about keeping everything on the top-level and merely adding `solarsystem/jpl` in addition; for now, I added `jplsbdb` as `sbdb` to `solarsystem/jpl` in form of a dynamic link (this required the `astroquery` imports in `jplsbdb/core.py` to be absolute instead of relative. Is this how you imagined it? This allows you to do both `from astroquery.jplsbdb import SBDB` and `from astroquery.solarsystem.jpl import SBDB`.
* what bugs me a little bit is the documentation: e.g., the API for this feature will either have to appear as `astroquery.jplsbdb` or `astroquery.solarsystem.jpl.sbdb`; I think it is not possible to have the same API/documentation for two different submodules - unless we copy, e.g.,  `docs/jplsbdb/jplsbdb.rst`, into a different location. I guess that this means that we can only have documentation for the top-level or the `solarsystem/jpl` version of `sbdb`. If that is the case, I would propose to go with `solarsystem/jpl/sbdb`.

Any advice or comments are welcome!

**Comment by keflavich** [comment]
Since users are generally going to access the docs from the index on the frontpage or from google, does it matter where they live?

Non-relative imports sound like a bad idea.  What if, instead, we just have `astroquery/solarsystem/jpl/sbdb/__init__.py` read `from .....jpl.core import SBDB, SBDBClass`, or something like that?

**Comment by mommermi** [comment]
Yes, `from ....jplsbdb import SBDB, SBDBClass` sounds like the way to go as it doesn't require dynamic links. Thanks! New commit in a few minutes...

**Comment by mommermi** [comment]
I redid the structure but I keep running into a problem. The `solarsystem.jpl.sbdb` API is not built properly. Sphinx keeps telling me `py:class reference target not found: astroquery.solarsystem.jpl.sbdb.SBDBClass` etc. I also realized that `solarsystem` is not fetched by `[automodsumm]`. Not sure why this happens.

Any thougts?

**Comment by bsipocz** [comment]
yes, we basically need to decide where to generate the API docs for these classes that are available at two namespaces. 

I would vote for dropping them at the convenience layer level?

**Comment by mommermi** [comment]
I followed your suggestion to drop the convenience layer docs and that magically solved my problem - seems like the way to go. I also added `jplhorizons` to `solarsystem/jpl`. I'm pretty happy with the result!

**Comment by mommermi** [comment]
anybody an idea what might cause this fail? https://travis-ci.org/astropy/astroquery/jobs/413720301

**Comment by pllim** [comment]
Yes, it is a bug with pytest 3.7.1. Pinning pytest to 3.7 should fix the issue until 3.7.2 is out.

**Comment by mommermi** [comment]
@pllim Thanks for the hint, but it is now failing with a different error message:

______________________________ ERROR collecting  _______________________________
/home/travis/miniconda/envs/test/lib/python3.6/site-packages/_pytest/runner.py:201: in __init__
    self.result = func()
/home/travis/miniconda/envs/test/lib/python3.6/site-packages/_pytest/runner.py:261: in <lambda>
    call = CallInfo(lambda: list(collector.collect()), "collect")
/home/travis/miniconda/envs/test/lib/python3.6/site-packages/_pytest/main.py:475: in collect
    for x in self._collect(arg):
/home/travis/miniconda/envs/test/lib/python3.6/site-packages/_pytest/main.py:519: in _collect
    for x in root._collectfile(pkginit):
E   AttributeError: 'DocTestModulePlus' object has no attribute '_collectfile'
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.98 seconds ============================

Any ideas on this one?

**Comment by pllim** [comment]
I think that was a different bug with 3.7. 😅  How about `pytest` 3.6?

File: .travis.yml [hunk 9]
```diff
@@ -43,7 +43,7 @@ env:
         - CONDA_DEPENDENCIES='requests beautifulsoup4 html5lib keyring aplpy matplotlib pyregion jinja2 flask regions'
         # pandas is used as a workaround of an astropy issue, sometimes it's required for the remote tests
         - CONDA_DEPENDENCIES_REMOTE='requests beautifulsoup4 html5lib keyring aplpy matplotlib pyregion jinja2 flask pandas regions'
-        - PIP_DEPENDENCIES='https://github.com/keflavich/httpbin/archive/master.zip astropy_healpix mocpy pytest-dependency'
+        - PIP_DEPENDENCIES='https://github.com/keflavich/httpbin/archive/master.zip astropy_healpix mocpy pytest==3.6 pytest-dependency'
```
Comment (bsipocz) [inline .travis.yml]: this has been dealt with upstream in `ci-helpers`, so please remove the change.

Comment (mommermi) [reply .travis.yml]: done

**Comment by mommermi** [comment]
I will have to let this rest until the end of August. I will try to fix all the test issues as soon as possible after that.

**Comment by mommermi** [comment]
I am back working on this. It seems that ``numpy.genfromtxt`` acts up for some earlier versions of python and numpy. I will have to write some additional functions that are used instead of ``genfromtxt`` if one of those older versions is used.

**Comment by mommermi** [comment]
@bsipocz @keflavich I think is finally done and ready to be merged.

**Comment by mommermi** [comment]
@keflavich @bsipocz Is there a chance that you could merge this this week? I would like to present it at a workshop next week.

**Comment by keflavich** [comment]
Yes, assuming it passes.

**Comment by keflavich** [comment]
Nevermind, it did pass, but github is showing me yellow incorrectly.

**Comment by mommermi** [comment]
Thanks!

**Comment by keflavich** [comment]
I added a changelog entry manually

**Comment by mommermi** [comment]
perfect, thanks!

**Pull request merged**